clinica/iotools/converters/genfi_to_bids/genfi_to_bids_utils.py

```diff
@@ -0,0 +1,650 @@
+    df: DataFrame
+        Dataframe without runs.
+
+    Returns
+    -------
+    DataFrame
+        Dataframe containing the correct run for each acquisition.
+    """
+    filter = ["source_id", "source_ses_id", "suffix", "dir_num"]
+    df1 = df[filter].groupby(filter).min()
+    df2 = df[filter].groupby(filter[:-1]).min()
+    df1 = df1.join(df2.rename(columns={"dir_num": "run_01_dir_num"}))
+    df_alt = df1.reset_index().assign(run=lambda x: (x.run_01_dir_num != x.dir_num))
+    return df_alt.assign(run_num=lambda df: df.run.apply(lambda x: f"run-0{int(x)+1}"))
+
+
+def complete_imaging_data(df_dicom: DataFrame) -> DataFrame:
+    """This function uses the raw information extracted from the images, to obtain all the information necessary for the BIDS
+
+    Parameters
+    ----------
+    df_dicom: DataFrame
+        Dataframe containing the data extracted from the images
+
+    Returns
+    -------
+    df_sub_ses_run: DataFrame
+        Dataframe with the data necessary for the BIDS
+    """
+
+    df_dicom = df_dicom.assign(
+        source_id=lambda df: df.source.apply(
+            lambda x: Path(Path(Path(x).parent).parent).name.split("-")[0]
+        )
+    )
+    df_dicom = df_dicom.assign(
+        source_ses_id=lambda df: df.source.apply(
+            lambda x: Path(Path(Path(x).parent).parent).name.split("-")[1]
+        )
+    )
+    df_dicom = df_dicom.assign(
+        origin=lambda df: df.source.apply(
+            lambda x: Path(Path(Path(Path(Path(x).parent).parent).parent).parent)
+        )
+    )
+
+    df_dicom = df_dicom.reset_index()
+
+    df_dicom = df_dicom.assign(source_ses_id=lambda df: df.source_ses_id.astype("int"))
+    df_dicom = df_dicom.assign(
+        genfi_version=lambda df: df.source_ses_id.apply(lambda x: f"GENFI{len(str(x))}")
+    )
+    df_1 = compute_baseline_date(df_dicom)
+
+    df_1 = compute_session_numbers(df_1)
+
+    df_sub_ses = df_dicom.merge(
+        df_1[["baseline", "session_id"]], how="inner", on=["source_id", "source_ses_id"]
+    )
+    df_sub_ses = df_sub_ses.assign(
+        participant_id=lambda df: df.source_id.apply(lambda x: f"sub-{x}")
+    )
+    df_sub_ses = compute_modality(df_sub_ses)
+
+    # take into account fieldmaps -> same method as for baseline (extract number of directory)
+    df_fmap = df_sub_ses.assign(
+        dir_num=lambda x: x.source.apply(lambda y: int(Path(Path(y).parent).name))
+    )
+
+    df_suf = merge_fieldmaps(df_fmap, identify_fieldmaps(df_fmap))
+
+    # take into account runs -> same method as for baseline (extract number of directory)
+    df_suf_dir = df_suf.assign(
+        dir_num=lambda df: df.source.apply(lambda x: int(Path(Path(x).parent).name))
+    )
+    df_alt = compute_runs(df_suf_dir)
+    df_sub_ses_run = df_suf_dir.merge(
+        df_alt[["source_id", "source_ses_id", "suffix", "dir_num", "run_num"]],
+        how="left",
+        on=["source_id", "source_ses_id", "suffix", "dir_num"],
+    )
+
+    ##builds path to bids
```

Comment:
Could be made a bit easier to read I think.
Maybe something like this ? (haven't tested though...)

```python
def compose_filename(df: DataFrame, cols: List[str], joiner: str) -> str:
    return joiner.join([df[col] for col in cols])

df_sub_ses_run = df_sub_ses_run.assign(
    bids_filename=lambda df: compose_filename(
        df, ["participant_id", "session_id", "run_num", "suffix"], "_"
    ),
    bids_full_path=lambda df: compose_filename(
        df, ["participant_id", "session_id", "datatype", "bids_filename"], "/"
    )
)
```

Reply:
Actually, I just realized I wrote this too quickly and it doesn't work.
You should be able to simplify and do though:

```python
df_sub_ses_run = df_sub_ses_run.assign(
    bids_filename=lambda df: df[["participant_id", "session_id", "run_num", "suffix"]].agg("_".join, axis=1),
    bids_full_path=lambda df: df[["participant_id", "session_id", "datatype", "bids_filename"]].agg("/".join, axis=1)
)
```

Reply:
works ! done :)